chapter6_marl/marl_network.py

```diff
@@ -0,0 +1,244 @@
+        self.agent_num = agent_num
+        self.actor_critic_networks = nn.ModuleList(
+            [ActorCriticNetwork(obs_shape, action_shape) for _ in range(agent_num)]
+        )
+
+    # delimiter
+    def forward(self, local_obs: torch.Tensor) -> ttorch.Tensor:
+        """
+        **Overview**:
+            The computation graph of independent actor-critic network, serially processing each agent's ``local_obs`` and output
+            the cooresponding policy logit and value respectively.
+        """
+        return ttorch.cat([net(local_obs[:, i:i + 1]) for i, net in enumerate(self.actor_critic_networks)], dim=1)
+
+
+# delimiter
+class CTDEActorCriticNetwork(nn.Module):
+
+    def __init__(self, agent_num: int, local_obs_shape: int, global_obs_shape: int, action_shape: int) -> None:
+        """
+        **Overview**:
+            The definition of centralized training decentralized execution (CTDE) actor-critic network in policy gradient algorithms for multi-agent scenarios.
+            Each agent shares the same parameters in the network so that they can be processed as a batch in parallel.
+            The input of value network is ``global_obs`` while the input of policy network is ``local_obs``.
+            Global information used in value network can provide more guidance for the training of policy network.
+            Local information used in policy network can make the policy network more robust to the decentralized execution.
+        """
+        # PyTorch necessary requirements for extending ``nn.Module`` . Our network should also subclass this class.
+        super(CTDEActorCriticNetwork, self).__init__()
+        # Define local and global encoder respectively.
+        self.agent_num = agent_num
+        self.local_encoder = nn.Sequential(
+            nn.Linear(local_obs_shape, 32),
+            nn.ReLU(),
+            nn.Linear(32, 64),
+            nn.ReLU(),
+        )
+        self.global_encoder = nn.Sequential(
+            nn.Linear(global_obs_shape, 32),
+            nn.ReLU(),
+            nn.Linear(32, 64),
+            nn.ReLU(),
+        )
+        # Define discrete action logit output network, just one-layer FC.
+        self.policy_head = nn.Linear(64, action_shape)
+        # Define scalar value output network.
+        self.value_head = nn.Linear(64, 1)
+
+    # delimiter
+    def forward(self, local_obs: torch.Tensor, global_obs: torch.Tensor) -> ttorch.Tensor:
+        """
+        **Overview**:
+            The computation graph of CTDE actor-critic network, processing all agents' ``local_obs`` and ``global_obs`` and output
+            corresponding policy logit and value in parallel.
+            There are two possible designs for ``global_obs``: The former is a shared global state for all agents, i.e. $$(B, S)$$.
+            Tha latter is a kind of agent-specific global state, i.e. $$(B, A, S')$$.
```

Comment:
Maybe you can add this link https://di-engine-docs.readthedocs.io/zh_CN/latest/04_best_practice/marl_zh.html#id10, so that readers can learn more about why 'agent_specific_global_state' should be used.